File: services/_shared/src/yp_shared/ids.py
import ulid

def generate_id(prefix: str) -> str:
    """Generate a ULID with a given prefix."""
    return f"{prefix}_{ulid.new().str}"

def parse_id(prefixed_id: str) -> str:
    """Parse and validate a prefixed ULID, returning the raw ULID string."""
    try:
        prefix, ulid_str = prefixed_id.split("_", 1)
        # Validate that it's a correct ULID by parsing
        parsed = ulid.parse(ulid_str)
        return parsed.str
    except (ValueError, AttributeError) as e:
        raise ValueError(f"Invalid prefixed ID format: {prefixed_id}") from e

File: packages/shared/src/utils/ids.ts
import { ulid } from 'ulid';

/**
 * Generates a new ULID prefixed with the given string.
 */
export function generateId(prefix: string): string {
  return `${prefix}_${ulid()}`;
}

/**
 * Parses a prefixed ULID and returns the raw ULID string.
 * Throws an error if the format is invalid.
 */
export function parseId(prefixedId: string): string {
  if (!prefixedId || typeof prefixedId !== 'string') {
    throw new Error('Invalid ID format');
  }
  
  const parts = prefixedId.split('_');
  if (parts.length !== 2) {
    throw new Error(`Invalid prefixed ID format: ${prefixedId}`);
  }
  
  const [, ulidStr] = parts;
  if (!ulidStr || ulidStr.length !== 26) {
    throw new Error(`Invalid ULID length in prefixed ID: ${prefixedId}`);
  }
  
  return ulidStr;
}
